clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -42,8 +43,17 @@
     private int version;
     private long lastRefreshMs;
     private long lastSuccessfulRefreshMs;
-    private Cluster cluster;
+    /**
+     * KAFKA-3428: Cluster class is assumed to be safe for concurrent access. To allow high concurrency on fetch(), 
+     * which returns the pointer to cluster, this method is not synchronized and safety is assured by cluster being a volatile pointer.
+     */
+    private volatile Cluster cluster;
     private boolean needUpdate;
+    /**
+     * KAFKA-3428
+     * synchronization policy: topics is updated only within Metadata synchronized methods. The read queries to topics however
+     * are not protected by external locks and topics must use an implementation that is safe to do reads concurrent with updates.
+     */
```

Comment:
Instead of spreading these comments in many places, I would have a single comment on the line after the opening class brace explaining the general strategy. Also, I would move the two fields that are accessed without a lock into a separate group with a brief comment just before them.